tests/executequery_test.go

```diff
@@ -99,10 +174,1137 @@ func TestExecuteQuery_SingleSimpleRow(t *testing.T) {
+			)),
+		},
+	)
+	server.ExecuteQueryFn = mockExecuteQueryFn(nil,
+		&executeQueryAction{
+			response:    partialResultSet("token", structVal(strVal("foo"), intVal(100)), structVal(strVal("bar"), intVal(101))),
+			endOfStream: true,
+		})
+	// 2. Build the request to test proxy
+	req := testproxypb.ExecuteQueryRequest{
+		ClientId: t.Name(),
+		Request: &btpb.ExecuteQueryRequest{
+			InstanceName: instanceName,
+			Query:        "SELECT * FROM table",
+		},
+	}
+	// 3. Perform the operation via test proxy
+	res := doExecuteQueryOp(t, server, &req, nil)
+	// 4. Verify the read succeeds and gets the expected metadata & data
+	checkResultOkStatus(t, res)
+	assert.Equal(t, len(res.Metadata.Columns), 1)
+	assert.True(t, cmp.Equal(res.Metadata,
+		testProxyMd(column("struct",
+			structType(namelessStructField(strType()), namelessStructField(int64Type())))),
+		protocmp.Transform()))
+	assert.Equal(t, len(res.Rows), 2)
+	assertRowEqual(t, testProxyRow(structVal(strVal("foo"), intVal(100))), res.Rows[0], res.Metadata)
+	assertRowEqual(t, testProxyRow(structVal(strVal("bar"), intVal(101))), res.Rows[1], res.Metadata)
+}
+
+// Tests that a query runs successfully with struct metadata that has duplicate field names
+func TestExecuteQuery_StructWithDuplicateColumnNames(t *testing.T) {
+	// 1. Instantiate the mock server with struct metadata having duplicate field names
+	server := initMockServer(t)
+	server.PrepareQueryFn = mockPrepareQueryFn(nil,
+		&prepareQueryAction{
+			response: prepareResponse([]byte("foo"), md(
+				// Struct fields have the same name "foo"
+				column("struct", structType(structField("foo", strType()), structField("foo", int64Type()))),
+			)),
+		},
+	)
+	server.ExecuteQueryFn = mockExecuteQueryFn(nil,
+		&executeQueryAction{
+			response:    partialResultSet("token", structVal(strVal("foo"), intVal(100)), structVal(strVal("bar"), intVal(101))),
+			endOfStream: true,
+		})
+	// 2. Build the request to test proxy
+	req := testproxypb.ExecuteQueryRequest{
+		ClientId: t.Name(),
+		Request: &btpb.ExecuteQueryRequest{
+			InstanceName: instanceName,
+			Query:        "SELECT * FROM table",
+		},
+	}
+	// 3. Perform the operation via test proxy
+	res := doExecuteQueryOp(t, server, &req, nil)
+	// 4. Verify the read succeeds and gets the expected metadata & data
+	checkResultOkStatus(t, res)
+	assert.Equal(t, len(res.Metadata.Columns), 1)
+	assert.True(t, cmp.Equal(res.Metadata,
+		testProxyMd(column("struct",
+			structType(structField("foo", strType()), structField("foo", int64Type())))),
+		protocmp.Transform()))
+	assert.Equal(t, len(res.Rows), 2)
+	assertRowEqual(t, testProxyRow(structVal(strVal("foo"), intVal(100))), res.Rows[0], res.Metadata)
+	assertRowEqual(t, testProxyRow(structVal(strVal("bar"), intVal(101))), res.Rows[1], res.Metadata)
+}
+
+// Tests that the operation fails if the final successful stream response lacks a resume token
```

Comment:
Does server always returns a resume token even in the end?

Reply:
Yes